source/adapters/level_zero/kernel.cpp

```diff
@@ -191,34 +191,62 @@ UR_APIEXPORT ur_result_t UR_APICALL urEnqueueKernelLaunch(
   ZE2UR_CALL(zeKernelSetGroupSize, (ZeKernel, WG[0], WG[1], WG[2]));
 
   bool UseCopyEngine = false;
+  bool LastCommandEventIncluded = false;
   _ur_ze_event_list_t TmpWaitList;
   UR_CALL(TmpWaitList.createAndRetainUrZeEventList(
-      NumEventsInWaitList, EventWaitList, Queue, UseCopyEngine));
+      NumEventsInWaitList, EventWaitList, Queue, UseCopyEngine,
+      &LastCommandEventIncluded));
 
   // Get a new command list to be used on this call
   ur_command_list_ptr_t CommandList{};
   UR_CALL(Queue->Context->getAvailableCommandList(
       Queue, CommandList, UseCopyEngine, true /* AllowBatching */));
 
+  uint32_t WaitListLength = TmpWaitList.Length;
+  uint32_t WaitEventOffset = 0;
+
+  if (Queue->Device->useDriverInOrderLists() && Queue->isInOrderQueue()) {
```

Comment:
why this heavy code is needed in launch function ?

Reply:
Will be updating/removing all this, as it's mostly just remnants from Jaime's old PR of this (https://github.com/intel/llvm/pull/11942). Discussed proper simple solution for this offline with @MichalMrozek. Will update shortly.